subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1038 @@
+
+	ctx->fds[1].fd = ctx->event_fd;
+	ctx->fds[1].events = POLLIN;
+
+	ctx->num_clients = 0;
+	ctx->infinite = 1;
+
+	return ctx->server_fd;
+}
+
+int accept_new_client(int server_fd)
+{
+	int new_socket;
+	socklen_t addrlen;
+	struct sockaddr_storage sa;
+
+	memset(&sa, 0, sizeof(sa));
+	addrlen = sizeof(sa);
+
+#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
+	new_socket = accept(server_fd, (struct sockaddr *)&sa, &addrlen);
+#else
+	new_socket = zsock_accept(server_fd, (struct sockaddr *)&sa, &addrlen);
+#endif
+	if (new_socket < 0) {
+		LOG_ERR("accept failed");
+		return new_socket;
+	}
+
+	return new_socket;
+}
+
+int http_server_start(struct http_server_ctx *ctx)
+{
+	LOG_INF("Waiting for incoming connections at http://%s:%d", SERVER_IPV4_ADDR, ctx->config.port);
+
+	eventfd_t value;
+
+	do {
+		int ret = poll(ctx->fds, ctx->num_clients + 2, 0);
+
+		if (ret < 0) {
+			LOG_ERR("poll failed");
+			return -errno;
+		}
+
+		for (int i = 0; i < ctx->num_clients + 2; i++) {
+			struct http_client_ctx *client = &ctx->clients[i - 2];
+
+			if (ctx->fds[i].revents & POLLERR) {
+				LOG_ERR("Error on fd %d", ctx->fds[i].fd);
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			if (ctx->fds[i].revents & POLLHUP) {
+				LOG_INF("Client on fd %d has disconnected", ctx->fds[i].fd);
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			if (!(ctx->fds[i].revents & POLLIN)) {
+				continue;
+			}
+
+			if (i == 0) {
+				int new_socket = accept_new_client(ctx->server_fd);
+
+				bool found_slot = false;
+
+				for (int j = 2; j < MAX_CLIENTS + 1; j++) {
+					if (ctx->fds[j].fd != 0) {
+						continue;
+					}
+
+					ctx->fds[j].fd = new_socket;
+					ctx->fds[j].events = POLLIN;
+
+					initialize_client_ctx(&ctx->clients[j - 2], new_socket);
+
+					if (j > ctx->num_clients) {
+						ctx->num_clients++;
+					}
+
+					found_slot = true;
+					break;
+				}
+				if (!found_slot) {
+					LOG_INF("No free slot found.");
+#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
+					close(new_socket);
+#else
+					zsock_close(new_socket);
+#endif
+				}
+				continue;
+			}
+
+			if (i == 1) {
+#ifdef __ZEPHYR__
+				eventfd_read(ctx->event_fd, &value);
+#else
+				read(ctx->event_fd, &value, sizeof(value));
+#endif
+				LOG_DBG("Received stop event. exiting ..");
+
+				return 0;
+			}
+
+#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
+			int valread = recv(client->client_fd, client->buffer + client->offset,
+					   sizeof(client->buffer) - client->offset, 0);
+#else
+			int valread = zsock_recv(client->client_fd, client->buffer + client->offset,
+						 sizeof(client->buffer) - client->offset, 0);
+#endif
+
+			if (valread <= 0) {
+				if (valread == 0) {
+					LOG_INF("Connection closed by peer");
+				} else {
+					LOG_ERR("ERROR reading from socket");
+				}
+
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			client->offset += valread;
+			handle_http_request(ctx, client, i);
```

Comment:
We need to verify the result of `handle_http_request()` here - in case of any parsing errors, the connection with the client should be closed. With exception to `-EAGAIN` which we agreed to be a special code to indicate that more data is expected by the parser.